### PR TITLE
Removing model part pointer

### PR DESCRIPTION
**Pull request by RiccardoRossi**
this commit is preparatory to the big change of the Model (see #2417 )

it removes the defintion of ModelPart::Pointer and ensures that ModelPart::Pointers are NOT used anywhere in the code. 

It also modifies the python interface of the model so that only references are given back by the function CreateSubModelPart and GetSubModelPart, thus ensuring that the exclusive owner of the submodelparts is their father mp.

### CODE REVIEW

**Comment by loumalouomega** [comment]
Maybe replacing ModelPart::Pointer for auto would simplify the refactoring, but maybe you can find that dangerous

**Comment by RiccardoRossi** [comment]
it is only an intermediate change. we will need additional changes for the model

**Comment by loumalouomega** [comment]
Well, the thing is that if we decide to change this again in the future, thanks to the auto, we won't need to do this again

File: applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp [hunk 1]
```diff
@@ -124,11 +124,11 @@ void SetMeshToInitialConfiguration(
 
 //******************************************************************************
 //******************************************************************************
-ModelPart::Pointer GenerateMeshPart(ModelPart &rModelPart,
+ModelPart* GenerateMeshPart(ModelPart &rModelPart,
                                     const std::string &rElementName) {
   KRATOS_TRY;
 
-  ModelPart::Pointer pmesh_model_part;
+  ModelPart* pmesh_model_part;
```
Comment (RiccardoRossi) [inline applications/MeshMovingApplication/custom_utilities/move_mesh_utilities.cpp]: mmm a note for myself: this should be removed

File: kratos/sources/model_part.cpp [hunk 2]
```diff
@@ -1092,37 +1092,28 @@ void ModelPart::RemoveConditionsFromAllLevels(Flags identifier_flag)
 }
 
 
-ModelPart::Pointer  ModelPart::CreateSubModelPart(std::string const& NewSubModelPartName)
+ModelPart*  ModelPart::CreateSubModelPart(std::string const& NewSubModelPartName)
 {
     if (mSubModelParts.find(NewSubModelPartName) == mSubModelParts.end())
     {
-        ModelPart::Pointer p_model_part(new ModelPart(NewSubModelPartName));
+        Kratos::shared_ptr<ModelPart>  p_model_part = Kratos::make_shared<ModelPart>(NewSubModelPartName);
```
Comment (RiccardoRossi) [inline kratos/sources/model_part.cpp:1099]: i would love to use something else here (a unique pointer would be best) but i cannot use it in the mSubModelParts

**Review by pooyan-dadvand — COMMENTED** [review]
@RiccardoRossi Thank you very much for your effort in separating the Pointer change from the rest of the PR. I have noticed that you have changed the CreateSubModelPart to return a raw pointer. Have you considered to return directly a reference? Then you may also use a unique pointer for storing the sup model part as you mentioned

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand we can definitely return a reference, it would actually allow me to get rid of the "chapuza" i mention in the modelpart python interface, however i would not do it in this PR since it is a major change and it will affect the syntax in many places. 

also as of now there is no way to store unique pointers in a SubModelParContainer (see issue #2198). Btw it is also not possible to store there raw pointers, or at least i was not able to...

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand finally i did modify the function to get back a reference

**Review by pooyan-dadvand — APPROVED** [review]
Let's go ahead! 👍

**Comment by pooyan-dadvand** [comment]
@RiccardoRossi asked me to merge it.

**Comment by josep-m-carbonell** [comment]
@pooyan-dadvand  after the changes of this merge Solid and Pfem applications crash with a segmentation fault. Tests also crash. @KratosMultiphysics/technical-committee  @KratosMultiphysics/pfem

**Comment by josep-m-carbonell** [comment]
here the valgrind error report:

==11730== 
==11730== Process terminating with default action of signal 11 (SIGSEGV)
==11730==  Access not within mapped region at address 0x100055FD80
==11730==    at 0x7FB2673: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<Kratos::ModelPart> >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<Kratos::ModelPart> > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_find_before_node(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long) const (hashtable.h:1426)
==11730==    by 0x7FB24EC: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<Kratos::ModelPart> >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<Kratos::ModelPart> > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_find_node(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long) const (hashtable.h:632)
==11730==    by 0x7FB2414: std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<Kratos::ModelPart> >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<Kratos::ModelPart> > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::find(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (hashtable.h:1307)
==11730==    by 0x7FB23AC: std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::shared_ptr<Kratos::ModelPart>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::shared_ptr<Kratos::ModelPart> > > >::find(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/cpuigbo/Kratos/libs/Kratos.cpython-35m-x86_64-linux-gnu.so)
==11730==    by 0x9696CFE: Kratos::PointerHashMapSet<Kratos::ModelPart, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, Kratos::ModelPart::GetModelPartName, std::shared_ptr<Kratos::ModelPart> >::find(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (in /home/cpuigbo/Kratos/libs/libKratosCore.so)
==11730==    by 0x9792118: Kratos::ModelPart::CreateSubModelPart(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (model_part.cpp:1097)
==11730==    by 0x803852F: pybind11::cpp_function::cpp_function<std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(std::shared_ptr<Kratos::ModelPart> (Kratos::ModelPart::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}::operator()(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const (in /home/cpuigbo/Kratos/libs/Kratos.cpython-35m-x86_64-linux-gnu.so)
==11730==    by 0x8038490: std::shared_ptr<Kratos::ModelPart> pybind11::detail::argument_loader<Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>::call_impl<std::shared_ptr<Kratos::ModelPart>, pybind11::cpp_function::cpp_function<std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(std::shared_ptr<Kratos::ModelPart> (Kratos::ModelPart::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}&, 0ul, 1ul, pybind11::detail::void_type>(pybind11::cpp_function::cpp_function<std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(std::shared_ptr<Kratos::ModelPart> (Kratos::ModelPart::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}&, std::integer_sequence<unsigned long, 0ul, 1ul>, pybind11::detail::void_type&&) (in /home/cpuigbo/Kratos/libs/Kratos.cpython-35m-x86_64-linux-gnu.so)
==11730==    by 0x8038167: _ZNO8pybind116detail15argument_loaderIJPN6Kratos9ModelPartERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE4callISt10shared_ptrIS3_ENS0_9void_typeERZNS_12cpp_functionC1ISG_S3_JSC_EJNS_4nameENS_9is_methodENS_7siblingEEEEMT0_FT_DpT1_EDpRKT2_EUlS4_SC_E_EENSt9enable_ifIXntsr3std7is_voidISO_EE5valueESO_E4typeEOT1_ (in /home/cpuigbo/Kratos/libs/Kratos.cpython-35m-x86_64-linux-gnu.so)
==11730==    by 0x8037F87: void pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(std::shared_ptr<Kratos::ModelPart> (Kratos::ModelPart::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}, std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(pybind11::cpp_function::initialize<std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(std::shared_ptr<Kratos::ModelPart> (Kratos::ModelPart::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}&&, std::shared_ptr<Kratos::ModelPart> (*)(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(pybind11::detail::function_call&)#1}::operator()(pybind11::detail::function_call) const (pybind11.h:155)
==11730==    by 0x8037E87: void pybind11::cpp_function::initialize<pybind11::cpp_function::initialize<std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(std::shared_ptr<Kratos::ModelPart> (Kratos::ModelPart::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}, std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(pybind11::cpp_function::initialize<std::shared_ptr<Kratos::ModelPart>, Kratos::ModelPart, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling>(std::shared_ptr<Kratos::ModelPart> (Kratos::ModelPart::*)(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}&&, std::shared_ptr<Kratos::ModelPart> (*)(Kratos::ModelPart*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::{lambda(pybind11::detail::function_call&)#1}::__invoke(pybind11::detail::function_call) (pybind11.h:132)
==11730==    by 0x772E972: pybind11::cpp_function::dispatcher(_object*, _object*, _object*) (pybind11.h:619)

**Comment by josep-m-carbonell** [comment]
FALSE ALARM, after some recompilation the segmentation fault disappeared, I could not reproduce it again. Which is good. Sorry for the inconviencies
@pooyan-dadvand @KratosMultiphysics/technical-committee @KratosMultiphysics/pfem

**Comment by pooyan-dadvand** [comment]
Good to know!

Thanks for update!

**Pull request merged**